magicctapipe/scripts/lst1_magic/lst_m1_m2_cloud_correction.py

```diff
@@ -0,0 +1,417 @@
+    dHc : float
+        Cloud thickness
+    trans : float
+        Transmission of the cloud
+
+    Returns
+    -------
+    float
+        Height of a cloud transmission
+    """
+    t = pow(trans, ((x - Hc) / dHc).to_value(""))
+    t = np.where(x < Hc, 1, t)
+    t = np.where(x > Hc + dHc, trans, t)
+    return t
+
+
+def process_telescope_data(
+    input_file, output_file, telid, focal_eq, focal_eff, camgeom
+):
+    """
+    Corrects LST-1 and MAGIC data affected by a cloud presence
+
+    Parameters
+    ----------
+    input_file : str
+        Path to an input .h5 DL1 file
+    output_file : str
+        Path to a directory where to save an output corrected DL1 data file
+    telid : int
+        LST-1 and MAGIC telescope ids
+    focal_eq : float
+        Equivalent focal length
+    focal_eff : float
+        Effective focal length
+    camgeom : int
+        Camera geometry
+
+    Returns
+    -------
+    pandas.core.frame.DataFrame
+        Data frame of corrected DL1 parameters
+    """
+
+    Hc = 5900 * u.m  # cloud above the telescope
+    dHc = 1320 * u.m
+
+    data = []
+    dl1_params = read_table(input_file, "/events/parameters")
+    dl1_images = read_table(input_file, "/events/dl1/image_" + str(telid))
+
+    inds = np.where(
+        np.logical_and(dl1_params["intensity"] > 0.0, dl1_params["tel_id"] == telid)
+    )[0]
+    for index in inds:
+        event_id_lst = dl1_params["event_id_lst"][index]
+        obs_id_lst = dl1_params["obs_id_lst"][index]
+        event_id = dl1_params["event_id_magic"][index]
+        obs_id = dl1_params["obs_id_magic"][index]
+        if telid == 1:
+            event_id, obs_id = event_id_lst, obs_id_lst
+
+        pointing_az = dl1_params["pointing_az"][index]
+        pointing_alt = dl1_params["pointing_alt"][index]
+        pointing_alt_deg = np.rad2deg(pointing_alt)
+        zenith = 90.0 - pointing_alt_deg
+
+        x = dl1_params["x"][index] * u.m
+        y = dl1_params["y"][index] * u.m
+        psi = dl1_params["psi"][index] * u.deg
+        wl = dl1_params["width"][index] / dl1_params["length"][index]
+        log_intensity = np.log10(dl1_params["intensity"])[index]
+        timestamp = dl1_params["timestamp"][index]
+        time_diff = dl1_params["time_diff"][index]
+        n_islands = dl1_params["n_islands"][index]
+        n_pixels = dl1_params["n_pixels"][index]
+        signal_pixels = dl1_params["n_pixels"][index]
+
+        trans = 0.58  # transmission of the cloud
+        trans = trans ** (1 / np.cos(zenith))  # .to_value("")
+        nlayers = 10  # to simplify calculations with the model cloud is considered to be composed of multiple layers
+        Hcl = np.linspace(Hc, Hc + dHc, nlayers)  # position of each layer
+        transl = trans_height(Hcl, Hc, dHc, trans)  # transmissions of each layer
+        transl = np.append(
+            transl, transl[-1]
+        )  # add one more element for heights above the top of the cloud
```

Comment:
all those calculations of the layers of the cloud are generic and can you do it once before events loop to save time

Reply:
I'm reading calculating zenith for each event with pointing alt parameter from the table:
pointing_alt = dl1_params['pointing_alt'][index]
zenith = 90.0 - np.rad2deg(pointing_alt)
so this part needs to be in the loop.